pkg/agent/discoverycache.go

```diff
@@ -273,3 +276,98 @@ func (j *discoveryCache) updateCategoryCache() {
 		}
 	}
 }
+
+func (j *discoveryCache) updateManagedApplicationCache() {
+	log.Trace("updating managed application cache")
+
+	// Update cache with published resources
+	// TODO - Remove custom subresource and include subject subresource when added to model
+	existingManagedApplications := make(map[string]bool)
+	query := map[string]string{
+		apic.FieldsKey: apiServerFields + ",x-marketplace-subject",
+	}
+
+	managedApps, _ := GetCentralClient().GetAPIV1ResourceInstancesWithPageSize(
+		query, agent.cfg.GetEnvironmentURL()+"/managedapplications", apiServerPageSize,
+	)
+
+	for _, managedApp := range managedApps {
+		agent.cacheManager.AddManagedApplication(managedApp)
+		existingManagedApplications[managedApp.Metadata.ID] = true
+	}
+
+	if j.refreshAll {
+		// Remove managed applications that no longer exist
+		cacheKeys := agent.cacheManager.GetManagedApplicationCacheKeys()
+		for _, key := range cacheKeys {
+			if _, ok := existingManagedApplications[key]; !ok {
+				agent.cacheManager.DeleteManagedApplication(key)
+			}
+		}
+	}
+}
+
+func (j *discoveryCache) updateAccessRequestCache() {
+	log.Trace("updating access request cache")
+
+	// Update cache with published resources
+	// TODO - Remove custom subresource and include references
+	existingAccessRequests := make(map[string]bool)
+	query := map[string]string{
+		apic.FieldsKey: apiServerFields + ",spec,x-marketplace-subscription",
+	}
+
+	accessRequests, _ := GetCentralClient().GetAPIV1ResourceInstancesWithPageSize(
+		query, agent.cfg.GetEnvironmentURL()+"/accessrequests", apiServerPageSize,
+	)
+
+	for _, accessRequest := range accessRequests {
+		ar := &mv1.AccessRequest{}
+		ar.FromInstance(accessRequest)
+		agent.cacheManager.AddAccessRequest(ar)
+		existingAccessRequests[accessRequest.Metadata.ID] = true
+		j.addSubscription(ar)
+	}
+
+	if j.refreshAll {
+		// Remove access requests that no longer exist
+		cacheKeys := agent.cacheManager.GetAccessRequestCacheKeys()
+		for _, key := range cacheKeys {
+			if _, ok := existingAccessRequests[key]; !ok {
+				agent.cacheManager.DeleteAccessRequest(key)
+			}
+		}
+	}
+}
+
+func (j *discoveryCache) addSubscription(ar *mv1.AccessRequest) {
+	// TODO - Use subscription reference subresource on AccessRequest instead of custom subresource
+	// once controller starts to populate it.
+	ars := ar.GetSubResource("x-marketplace-subscription")
+	if ars != nil {
+		arSubscription := ars.(map[string]interface{})
+		propertyVal := arSubscription["name"]
```

Comment:
A helper func for x-marketplace-subscription might be helpful like there is in the util package for x-agent-details to reduce the type conversions.

Reply:
Sure, will add it

Reply:
Done